tests_e2e/tests/lib/add_network_security_group.py

```diff
@@ -0,0 +1,140 @@
+                        "properties": {{
+                            "description": "Allows inbound SSH connections from the orchestrator machine.",
+                            "protocol": "Tcp",
+                            "sourcePortRange": "*",
+                            "destinationPortRange": "22",
+                            "sourceAddressPrefix": "{self._my_ip_address}",
+                            "destinationAddressPrefix": "*",
+                            "access": "Allow",
+                            "priority": 100,
+                            "direction": "Inbound"
+                        }}
+                    }}
+                ]
+            }}
+        }}"""))
+
+        #
+        # Add reference to the NSG to the properties of the subnets.
+        #
+        # The subnets are a copy property of the virtual network in LISA's ARM template:
+        #
+        #     {
+        #         "condition": "[empty(parameters('virtual_network_resource_group'))]",
+        #         "apiVersion": "2020-05-01",
+        #         "type": "Microsoft.Network/virtualNetworks",
+        #         "name": "[parameters('virtual_network_name')]",
+        #         "location": "[parameters('location')]",
+        #         "properties": {
+        #             "addressSpace": {
+        #                 "addressPrefixes": [
+        #                     "10.0.0.0/16"
+        #                 ]
+        #             },
+        #             "copy": [
+        #                 {
+        #                     "name": "subnets",
+        #                     "count": "[parameters('subnet_count')]",
+        #                     "input": {
+        #                         "name": "[concat(parameters('subnet_prefix'), copyIndex('subnets'))]",
+        #                         "properties": {
+        #                             "addressPrefix": "[concat('10.0.', copyIndex('subnets'), '.0/24')]"
+        #                         }
+        #                     }
+        #                 }
+        #             ]
+        #         }
+        #     }
+        #
+        network_resource = self._get_resource(resources, "Microsoft.Network/virtualNetworks")
+
+        # Add a dependency on the NSG
+        nsg_reference = f"[resourceId('Microsoft.Network/networkSecurityGroups', '{NETWORK_SECURITY_GROUP}')]"
+        network_resource_dependencies = network_resource.get("dependsOn")
+        if network_resource_dependencies is None:
+            network_resource["dependsOn"] = [nsg_reference]
+        else:
+            network_resource_dependencies.append(nsg_reference)
+
+        subnets_copy = network_resource["properties"].get("copy") if network_resource.get("properties") is not None else None
+        if subnets_copy is None:
+            raise Exception("Cannot find the copy property of the virtual network in the ARM template")
+
+        subnets = [i for i in subnets_copy if i["name"] == 'subnets']
```

Comment:
I would check with get, incase key(name) is missing. Otherwise, we may get keynotfound error

Reply:
yup, thanks